### ISSUE TITLE
Group the services list by categories

### ISSUE THREAD

**Issue by AlexMartinFR** [tighten/takeout#88]
## Issue:

Takeout currently sorts the services list by alphabetical order.
As more and more services will be added to the list, it might become a bit confusing to browse.

## Suggestion:

Each service could have a `category` property, which could be used to at least group similar services.

Examples of categories:
- database
- storage
- search
- cache
- mail

## Improvement:

The`takeout enable` command could now have two steps instead of one:

- Select a category of services
- Choose the relevant service in the curated list

**Comment by mattstauffer** [comment]
I don't know if I want extra steps, but I do like the idea of grouping.

I wonder if there's a way we could build a faster UI or a UI with more information available without adding an extra step?

**Comment by AlexMartinFR** [comment]
I think [cli-menu's separators](https://github.com/php-school/cli-menu/blob/master/README.md) could be a good start:

![image](https://user-images.githubusercontent.com/19224681/91741874-9fc02880-ebb5-11ea-8b00-b4abbca68c5d.png)


Extra steps would only be necessary when the service list begins to be too long. The same way multi-steps forms are only relevant in some cases.

**Comment by mattstauffer** [comment]
Yah, I like that!

**Comment by AlexMartinFR** [comment]
Bear with me here, as this would need serious refactoring and is for now only generating the menu, but here is the raw idea:

```php
public function handle(): void
{
    […]

    $databaseServices = [
        'mariadb' => 'MariaDB',
        'mysql' => 'MySQL',
        'mssql' => 'MS SQL'
    ];

    $searchServices = [
        'meilisearch' => 'MeiliSearch',
        'elasticsearch' => 'ElasticSearch'
    ];

    $option = $this->menu('Services to enable')->setTitleSeparator('=');

    /* displayCategory() private method */
    $option->addStaticItem('Database:')->addStaticItem('---------');

    foreach ($databaseServices as $service => $name) {
        $option->addOption($service, $name);
    }

    $option->addLineBreak('', 1);
    /* End of the displayCategory() private method */

    /* Start of the displayCategory() private method */
    $option->addStaticItem('Search:')->addStaticItem('-------');

    foreach ($searchServices as $service => $name) {
        $option->addOption($service, $name);
    }

    $option->addLineBreak('', 1);
    /* End of the displayCategory() private method */

    $option->setPadding(2, 5)->open();

    if (!$option) {
        return;
    }

    $this->enable($option);
}
```

## Terminal preview:

![image](https://user-images.githubusercontent.com/19224681/91745012-9d140200-ebba-11ea-8e2f-7caea76338eb.png)

**Comment by jonsugar** [comment]
Yes, please! 

In the current project to port [Lambo](https://github.com/tighten/lambo) to PHP, I am experimenting with calling takeout so that I can find enabled databases and prompt the user to start them if required (for example, when configuring a project with a Jetstream UI). 

At the moment I am calling `takeout list --json` then collecting and filtering the output. I had mentioned to Matt in a telegram chat:

>Would be nice if something like the following was possible:
>
> `takeout list --json --stopped --type=database`

**Comment by mattstauffer** [comment]
I still think this is a great idea, and I love Jon's note about using these in the CLI

**Comment by AlexMartinFR** [comment]
As soon as Takeout integrate a categories concept for its services, I'll gladly leverage it to submit that PR 🙂

I've tried to add the concept of categories to the core of Takeout, but I am not comfortable enough with the codebase.
I hope someone else will be able to contribute that!

**Comment by danielebarbaro** [comment]
takeout use `nunomaduro/laravel-console-menu` package, a simple wrapper of `php-school/cli-menu` but there aren't subItem (https://github.com/php-school/cli-menu#sub-menu-item) or Split Item (https://github.com/php-school/cli-menu#split-item-1) implementations. In this case also Jon can use it

One way could be make a PR to `nunomaduro/laravel-console-menu` and add some new feature that we can use in takeout.

Alternative way is grouping services and iterate in some way 
```php
foreach($groups as $group => $values ) {
    $option = $this->menu("Group {$group}", $this->enableableServices($values))
            ->addLineBreak('', 1)
            ->setPadding(2, 5);
}
```

In all cases to a better classification/grouping of Service we can add a property on `BaseService`. This property is easy to group with something like `$collection->groupBy()`

what do you think about it? 
I know quite well 😎 the codebase and if you want i can handle this task.

**Comment by AlexMartinFR** [comment]
Grouping and iterating is the way to go IMHO 👍 

I what I envision, each service should have a `category` property.
Then, we'll only have to iterate through every categories to get the UI we want.
(I'm basically describing your code snippet here.)

If you can brings categories to every service that I can then read to sort everything, that would be very welcome indeed!
Thanks 😎

PS: That terminal preview is not just a mockup, it's actually a screenshot of a running hacked version of Takeout. I didn't need to tweak the `nunomaduro/laravel-console-menu` package to do so 🙂

![image](https://user-images.githubusercontent.com/19224681/94065471-79279300-fdeb-11ea-91a4-fcc80e6da17a.png)

**Comment by danielebarbaro** [comment]
Services grouped by category in #135 .
If you have questions, feel free to ask. 🤓

<img width="1533" alt="Screenshot 2020-09-24 at 12 31 02" src="https://user-images.githubusercontent.com/4376886/94134065-d4946800-fe61-11ea-8d99-50915490abcf.png">

I never used Minio maybe storage isn't the right place 🙈

**Comment by AlexMartinFR** [comment]
You rock, Daniele!

I've added a custom visual improvement I wanted to add, feel free to implement or reject it :)

https://github.com/tighten/takeout/pull/135/files#r494363191

**Comment by mattstauffer** [comment]
Closed via #135
